File: pa2-starter/pa2-starter/java/csc460/csps/CSP.java
package csc460.csps;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import csc460.csps.constraints.*;
import csc460.SearchState;
import csc460.searchproblems.SearchProblem;
import csc460.BoardCoordinate;

/**
 * Represents a constraint satisfaction problem (CSP). DomainType should be the type of the
 * values in the domain, e.g., for Sudoku, you might use Integer for this, while for a sheduling
 * problem, you would probably use a String representing the name of the entity being scheduled. 
 * 
 * This class was modified to include the use of forward checking, least constraining value, and minimum remaining values. 
 * The methods getSuccessors, selectNextVariableIndex, orderValues, and forwardCheck were modified to include these features. 
 * 
 * Generative AI: Some methods in this class were modified with the use of generative AI. A section in the orderValues method
 * was produced by ChatGPT to help me correctly sort the orderedValues list. forwardCheck was in large part produced by
 * Github Copilot, and further modifications were made by me to ensure that the method was correctly implemented.
 * 
 * @author Hank Feild
 * @author Bradford Torpey
 */
public abstract class CSP<DomainType> implements SearchProblem {
    protected SearchState startState;
    protected ArrayList<String> variables;
    protected ArrayList<DomainType> domain;
    protected ArrayList<Constraint> constraints;
    protected HashMap<String, Integer> variableIndexLookup;
    protected boolean useForwardChecking, useLeastConstrainingValue, useMinimumRemainingValues;


    /**
     * Represents a CSP state, which is a list of assignments and domains that
     * correspond to the variable ordering (see the outer CSP class). The value
     * of unassigned variables is null.
     */
    public class CSPState extends SearchState {
        private ArrayList<DomainType> assignments;
        private ArrayList<ArrayList<DomainType>> domains;

        /**
         * Initializes the state.
         * 
         */
        public CSPState(ArrayList<DomainType> assignments, ArrayList<ArrayList<DomainType>> domains){
            this.assignments = assignments;
            this.domains = domains;
        }

        /**
         * @return The current set of variable assignments.
         */
        public ArrayList<DomainType> getAssignments(){
            return assignments;
        }

        /**
         * @return The current set of variable domains.
         */
        public ArrayList<ArrayList<DomainType>> getDomains(){
            return domains;
        }

        /**
         * @return A deep copy of this state.
         */
        public CSPState clone(){
            ArrayList<DomainType> clonedAssignments = new ArrayList<DomainType>(assignments);
            ArrayList<ArrayList<DomainType>> clonedDomains = new ArrayList<ArrayList<DomainType>>();
            for(ArrayList<DomainType> domain : domains){
                clonedDomains.add(new ArrayList<DomainType>(domain));
            }
            return new CSPState(clonedAssignments, clonedDomains);
        }

        /**
         * Unused.
         */
        @Override
        public BoardCoordinate getAgentCoordinates() {
            return null;
        }

        /**
         * Unused.
         */
        @Override
        public double getCost() {
            return 0;
        }

        /**
         * Unused.
         */
        @Override
        public double getDistance() {
            return 0;
        }

        /**
         * Unused.
         */
        @Override
        public String getAction() {
            return null;
        }

        /**
         * @returns The assignments as a string.
         */
        public String toString(){
            return assignments.toString();
        }
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    /**
     * Initializes an empty CSP.
     */
    public CSP(){
        this(false, false, false);
    }

    /**
     * Initializes an empty CSP.
     * 
     * @param useForwardChecking Set to true to use forward checking on successors.
     * @param useLeastConstrainingValue Set to true to use the least constraining value for domain value selection.
     * @param useMinimumRemainingValues Set to true to use minimum remaining values for variable selection.
     */
    public CSP(boolean useForwardChecking, boolean useLeastConstrainingValue, boolean useMinimumRemainingValues){
        this.variables = new ArrayList<String>();
        this.domain = new ArrayList<DomainType>();
        this.constraints = new ArrayList<Constraint>();
        this.variableIndexLookup = new HashMap<String, Integer>();
        this.useForwardChecking = useForwardChecking;
        this.useLeastConstrainingValue = useLeastConstrainingValue;
        this.useMinimumRemainingValues = useMinimumRemainingValues;
    }

    /**
     * Initializes the CSP. If there are initial assignments, get the starting
     * state after the constructor has been called and add the assignments
     * there.
     *
     * @param variables A list of variable names.
     * @param domain A list of domain values; all variables are initially given the same domain.
     * @param constraints A list of constraints that define the problem.
     * @param useForwardChecking Set to true to use forward checking on successors.
     * @param useLeastConstrainingValue Set to true to use the least constraining value for domain value selection.
     * @param useMinimumRemainingValues Set to true to use minimum remaining values for variable selection.
     */
    public CSP(ArrayList<String> variables, ArrayList<DomainType> domain, ArrayList<Constraint> constraints,
               boolean useForwardChecking, boolean useLeastConstrainingValue, boolean useMinimumRemainingValues){
        this.variables = variables;
        this.domain = domain;
        this.constraints = constraints;
        this.variableIndexLookup = new HashMap<String, Integer>();
        this.useForwardChecking = useForwardChecking;
        this.useLeastConstrainingValue = useLeastConstrainingValue;
        this.useMinimumRemainingValues = useMinimumRemainingValues;
        ArrayList<ArrayList<DomainType>> domains = new ArrayList<ArrayList<DomainType>>();
        ArrayList<DomainType> initialAssignments = new ArrayList<DomainType>();

        // Set each variables list index, domain, and initial assignment (null).
        for(int i = 0; i < variables.size(); i++){
            variableIndexLookup.put(variables.get(i), i);
            domains.add(new ArrayList<DomainType>(domain));
            initialAssignments.add(null);
        }

        startState = new CSPState(initialAssignments, domains);
    }

    /**
     * Implements backtracking; only possible assignment to the next variable
     * are emitted, and those that are must be consistent with previous
     * assignments.
     * 
     * @param state The state to find successors of.
     * @return A collection of successor states.
     */
    @Override
    public Iterable<SearchState> getSuccessors(SearchState state) {
        CSPState currentState = (CSPState) state;
        ArrayList<SearchState> successors = new ArrayList<SearchState>();

        ArrayList<DomainType> assignments = currentState.getAssignments();
        // If there are no missing assignments, there are no successors, so 
        // stop here.
        if(assignments.indexOf(null) < 0){
            return successors;
        }

        // Get the index of the next empty assignment.
        // int i = assignments.indexOf(null);
        int variableIndex = selectNextVariableIndex(currentState);

        // Add any consistent assignment to that variable to the list of
        // successors.
        for(DomainType value : orderValues(currentState, variableIndex)){
            CSPState successorState = currentState.clone();
            successorState.getAssignments().set(variableIndex, value);
            if(constraintsSatisfied(successorState.getAssignments())){
                if(useForwardChecking){ // Forward checking
                    successorState = forwardCheck(successorState, variableIndex, value); // Update successor state with forward checking
                    boolean hasEmptyDomain = false; 
                    for(DomainType domainValue : successorState.getDomains().get(variableIndex)){ // Check if any domain is empty
                        if(domainValue == null){ // If domain is empty, break and continue to next successor
                            hasEmptyDomain = true;
                            break;
                        }
                    }
                    if(hasEmptyDomain){ // If domain is empty, continue to next successor
                        continue;
                    }
                }

                successors.add(successorState); // Add successor to list of successors
            }
        }
        return successors;
    }

    /**
     * Selects the next value to assign to the variable at the given index. If useMinimumRemainingValues
     * is set to true for this CSP, the value with the fewest remaining values in the domain is selected.
     * Otherwise, the first unassigned variable is returned.
     * 
     * @param state The current state of the CSP.
     * 
     * @return The index of the next variable to assign a value to.
     */
    public int selectNextVariableIndex(CSPState state){
        if (useMinimumRemainingValues) { // Minimum remaining values
            int minRemainingValues = Integer.MAX_VALUE; // Initialize minRemainingValues to maximum value (2^31 - 1)
            int minRemainingValuesIndex = -1; // Initialize minRemainingValuesIndex to -1
            ArrayList<ArrayList<DomainType>> domains = state.getDomains(); // Get the domains from state
            for (int i = 0; i < domains.size(); i++) { // For each domain
                ArrayList<DomainType> domain = domains.get(i);
                if (state.getAssignments().get(i) == null && domain.size() < minRemainingValues) { // If the variable is unassigned and the domain size is less than minRemainingValues
                    minRemainingValues = domain.size(); // Update minRemainingValues
                    minRemainingValuesIndex = i; 
                }
            }
            return minRemainingValuesIndex; // Return the index of the variable with the minimum remaining values
        }
        return state.getAssignments().indexOf(null); 
    }

    /**
     * Returns an ordering of the values in the domain of the specified variable. If useLeastConstrainingValue
     * is set to true, the values are ordered by the number of values that would be eliminated from the domains
     * of other unassigned variables if the value were assigned to the variable at the given index. Otherwise, 
     * the initial ordering of the domain is returned.
     * 
     * @param state The current state of the CSP.
     * @param variableIndex The index of the variable to order the values of.
     * @return An ordered list of the values in the domain of the variable at the given index.
     */
    public ArrayList<DomainType> orderValues(CSPState state, int variableIndex){
        if(useLeastConstrainingValue){ // Least constraining value
            ArrayList<DomainType> domain = state.getDomains().get(variableIndex); // Get the domain of the variable at the given index
            ArrayList<Integer> removalCounts = new ArrayList<>(); // Initialize removalCounts to store # of values eliminated from domains of other unassigned variables
            for (DomainType value : domain) { // For each value in the domain
                int count = 0; // Reset count to 0
                for (int i = 0; i < state.getDomains().size(); i++) { 
                    if (i != variableIndex) { // If the index is not the variable index
                        CSPState newState = forwardCheck(state, variableIndex, value); // Get the new state with the value assigned to the variable at the given index
                        if (newState.getDomains().get(i).size() < state.getDomains().get(i).size()) { // If the domain size is less than the original domain size
                            count++; // Increment count
                        }
                    }
                }
                removalCounts.add(count); // Add count to removalCounts
            }
            // This section of code below was in part produced by ChatGPT to help me correctly sort the orderedValues list 
            // I had initially been unsure of how to use Collections.sort() and ChatGPT was able to help me understand how to use it correctly
            ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
            ArrayList<DomainType> orderedValues = new ArrayList<>(domain); // Initialize orderedValues to a copy of the domain
            orderedValues.sort((v1, v2) -> { // Sort orderedValues by the # of values eliminated from domains of other unassigned variables
                int index1 = domain.indexOf(v1); 
                int index2 = domain.indexOf(v2);
                return Integer.compare(removalCounts.get(index1), removalCounts.get(index2)); // Compare the # of values eliminated from the domains of other unassigned variables
            });
            return orderedValues; // Return orderedValues
            ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
        }
        return state.getDomains().get(variableIndex);
    } 
    /**
     * Returns a new state with the given variable assigned the given value, and the domains of the other variables updated.
     *
     * @param state The current state of the CSP.
     * @param variableIndex The index of the variable to assign the value to.
     * @param value The value to assign to the variable.
     * 
     * @return A new state with the given variable assigned the given value, and the domains of the other variables updated to
     *        be consistent with the new assignment.
     */
    public CSPState forwardCheck(CSPState state, int variableIndex, DomainType value) {
        // Deep copy.
        CSPState newState = state.clone();
        newState.getAssignments().set(variableIndex, value);

        // Convenience references
        ArrayList<DomainType> assignments = newState.getAssignments();
        ArrayList<ArrayList<DomainType>> domains = newState.getDomains();

        // This section of code below was produced by Github Copilot, and further modifications were made by me to ensure that the method was correctly implemented
        // I was unsure of how to correctly implement forward checking on some of the constraint types, and Github Copilot was able to help me understand how to implement it
        ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
        for (Constraint constraint : constraints) { // Iterate through constraints
            ArrayList<Integer> variableIndices = new ArrayList<>(); // Initialize variableIndices 
            for (Object variable : constraint.variables) { // For each variable in the constraint
                String variableString = (String) variable;
                Integer index = variableIndexLookup.get(variableString); // Get the index of the variable
                if (index != null) { // If the index is not null
                    variableIndices.add(index); // Add the index to variableIndices
                }
            }

            if (variableIndices.contains(variableIndex)) { // If variableIndices contains the variable index
                if (constraint instanceof AllDiffConstraint) { // AllDiffConstraint
                    for (int index : variableIndices) { // For each index in variableIndices
                        if (index != variableIndex) { // If the index is not the variable index
                            ArrayList<DomainType> domain = domains.get(index); 
                            domain.remove(value); // Remove the value from the domain
                        }
                    }
                } else if (constraint instanceof AllSameConstraint) { // AllSameConstraint
                    for (int index : variableIndices) { // For each index in variableIndices
                        if (index != variableIndex) { // If the index is not the variable index
                            ArrayList<DomainType> domain = domains.get(index);
                            domain.retainAll(Collections.singletonList(value)); // Retain only the value in the domain
                        }
                    }
                } else if (constraint instanceof MaxCountNConstraint) { // MaxCountNConstraint
                    MaxCountNConstraint maxCountConstraint = (MaxCountNConstraint) constraint; // Cast constraint to MaxCountNConstraint
                    int maxCount = maxCountConstraint.getMaxCount(); // Get the max count
                    int count = 0; // Initialize count to 0
                    for (int index : variableIndices) { // For each index in variableIndices
                        if (index != variableIndex) { // If the index is not the variable index
                            ArrayList<DomainType> domain = domains.get(index);
                            if (domain.contains(value)) { // If the domain contains the value
                                count++;
                                if (count > maxCount) { // If the count is greater than the max count
                                    domain.remove(value); // Remove the value from the domain
                                }
                            }
                        }
                    }
                }
            }
        }
        ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
        return newState;
    }


    /**
     * 
     * @param assignments
     * @return
     */
    public boolean constraintsSatisfied(ArrayList<DomainType> assignments){
        for(Constraint constraint : constraints){
            if(!constraint.isSatisfied(variableIndexLookup, assignments)){
                return false;
            }
        }
        return true;
    }

    /**
     * @param state The state to test.
     * @return True if the agent has reached the exit.
     */
    @Override
    public boolean isGoal(SearchState state) {
        ArrayList<DomainType> assignments = ((CSPState) state).getAssignments();
        return assignments.indexOf(null) < 0 && constraintsSatisfied(assignments);
    }

    /**
     * @return A string representation of variables and their assignments, as
     *         specified by the given search state. The format is:
     *         \t[variable]\t[value]
     */
    public String getAssignmentsAsString(SearchState state){
        ArrayList<DomainType> assignments = ((CSPState) state).getAssignments();
        StringBuilder sb = new StringBuilder();
        for(int i = 0; i < variables.size(); i++){
            sb.append('\t').append(variables.get(i))
              .append('\t').append(assignments.get(i))
              .append('\n');
        }

        return sb.toString();
    }

    /**
     * Unused.
     */
    @Override
    public void setHeuristic(String heuristic) {
    }

    /**
     * @return The starting state (the initial set of assignments).
     */
    @Override
    public SearchState getStartState() {
        return startState;
    }
    
    /**
     * This should be overridden.
     * @return A description of the format of the file that this class can read.
     *      This should be written so that it appears nicely on the command line.
     */
    public static String getFileFormatDescription(){
        return "This method should be overridden in the subclass; replace this message with a helpful CLI description.";
    }
}

File: pa2-starter/pa2-starter/python/csp.py
# Author: Hank Feild (hfeild@endicott.edu)

from math import sqrt
import sys
import problem 
import search
import random

class CSPState():
    def __init__(self, assignments, unassignedVariables, domains):
        '''
        Creates a CSPState out of the given assignments and per-variable domains.

        Parameters:
            assignments (dict[str]:Any): A dictionary of variable names to their assignments.
            unassignedVariables (list[str]): A list of unassigned variables.
            domains (dict[str]:list[Any]): A dictionary of variable names to their domains. Each domain is a list of possible values.
        '''
        self.assignments = assignments
        self.unassignedVariables = unassignedVariables
        self.domains = domains

    def __eq__(self, other):
        return str(self) == str(other)
    def __hash__(self):
        return hash(str(self))

    def __str__(self):
        return str((tuple(sorted(self.assignments.items())), tuple(sorted(self.unassignedVariables)), tuple(sorted(self.domains.items()))))

class CSP(problem.Problem):
    '''
    A class representing a CSP problem. Subclass this to define a specific CSP problem.
    '''

    # Keep this updated with all avaiable heuristics.
    heuristics = set()

    def __init__(self, cspConfigFilename=None, forwardChecking=False, minimumRemainingValues=False, leastConstrainingValue=False):
        '''
        Parameters:
            boardFilename (str): The filename of the board to load.
            heuristic (str): The heuristic to use for informed search.
        '''        
        self.forwardChecking = forwardChecking
        self.minimumRemainingValues = minimumRemainingValues
        self.leastConstrainingValue = leastConstrainingValue

        if cspConfigFilename == None:
            self.start = (None, None, None, None)
            self.assignments = {}
            self.domains = {}
            self.constraints = []
        else:
            self.loadFile(cspConfigFilename)


    def loadFile(self, cspConfigFilename):
        '''
        Parses a CSP config file. Should contain two or more lines:
            1. A list of variables, separated by spaces. Initially assigned values should be of the form variable=value.
            2. A list of domain values, space separated
            3. (0 or more lines) A list of constraints, one per line, each of the form 
                    CONSTRAINT_TYPE variable1 variable2 ... variableN
                - CONSTRAINT_TYPE is one of: 'AllDiff', 'AllSame', 'MaxCount n'
                - 'AllDiff' means all variables must have the same value
                - 'AllSame' means all variables must have different values
                - 'MaxCount n' means no more than n variables can have the same value; n is an integer

        Parameters:
            cspConfigFilename (str): The path to a file containing the CSP configuration.
        '''
        self.assignments = {}
        self.unassignedVariables = []
        self.domains = {}
        self.constraints = []
        with open(cspConfigFilename) as file:
            # Parse the variables.
            for variable in file.readline().rstrip().split():
                if '=' in variable:
                    self.assignments[variable.split('=')[0]] = variable.split('=')[1]
                else:
                    self.unassignedVariables.append(variable)
                    self.domains[variable] = []

            # Parse the domains.
            for domainValue in file.readline().rstrip().split():
                # Add each domain value to each unassigned variable's domain.
                for key in self.unassignedVariables:
                    self.domains[key].append(domainValue)

            # Parse the constraints.
            self.constraints = []
            for line in file:
                self.constraints.append(line.split())


        self.start = CSPState(self.assignments, self.unassignedVariables, self.domains)

    def successors(self, state):
        '''
        Produces a list of assignments to the next variable.

        Parameters:
            state ((assignments, unassignedVariables, domains)): The current state of the problem.

        Returns (list((str, (int,int), float, float))): A list of the
            states that can be reached from the given state. Each item is a
            4-tuple: (action, new state, cost of action, estimated cost to
            goal).
        '''
        successors = []
        
        if len(state.unassignedVariables) == 0:
            return successors
        
        variable = self.chooseVariable(state)

        for value in self.orderValues(state, variable):
            newAssignments = {k:v for k,v in state.assignments.items()}
            newAssignments[variable] = value
            newUnassignedVariables = state.unassignedVariables.copy()
            newUnassignedVariables.remove(variable)
            newDomains = {k:v.copy() for k,v in state.domains.items()}

            # Check constraints.
            if self.checkConstraints(newAssignments):
                newState = CSPState(newAssignments, newUnassignedVariables, newDomains)


                if self.forwardChecking:
                    newState = self.forwardCheck(newState, variable, value)

                    # If there are any empty domains, skip this successor.
                    # TODO: implement this check.

                    
                successors.append((variable + '=' + value, newState, 1, 0))
        
        return successors
    
    def chooseVariable(self, state):
        '''Selects the next unassigned variable to consider. If
        self.minimumRemainingValues is True, choose the variable with the fewest
        remaining values in its domain, otherwise an arbitrary unassigned variable
        will be selected.
        
        Parameters:
            state ((assignments, unassignedVariables, domains)): The current state of the problem.

        Returns (str): The name of the variable to assign next.
        '''
        if self.minimumRemainingValues:
            # TODO: implement minimum remaining values ordering; return the variable with the fewest remaining values in its domain.
            return state.unassignedVariables[0] # Replace this line!
        else:
            return state.unassignedVariables[0]
        
        

    def orderValues(self, state, variable):
        '''Orders the values of the given variable. If
        self.leastConstrainingValue is True, values are ordered by how many
        values they remmove from the remaining variables' domains, otherwise an
        an arbitrary ordering is selected.
        
        Parameters:
            state ((assignments, unassignedVariables, domains)): The current
            state of the problem. variable (str): The name of the variable to
            select a value for.

        Returns (list[Any]): The values to assign to the variable.
        '''
        if self.leastConstrainingValue:
            # TODO: implement least constraining value ordering; return the values in the domain of
            # the given variable, ordered by how many values they remove from the remaining variables' domains.
            # You will find the forwardCheck() method helpful here.
            return state.domains[variable] # Replace this line!
        else:
            return state.domains[variable]
    

    def checkConstraints(self, assignments, variable=None):
        '''
        Checks if the given assignment satisfies all constraints.

        Parameters:
            assignments (dict[str]:Any): The current assignment of variables.
            variable (str): Optional. If not None, only check the constraints that involve this variable.

        Returns (bool): True if the given assignment satisfies all constraints, False otherwise.
        '''
        for constraint in self.constraints:
            # Skip this constraint if necessary.
            if variable is not None and variable not in constraint[1:]:
                continue

            if constraint[0] == 'AllDiff':
                values = set()
                for var in constraint[1:]:
                    if var in assignments:
                        if assignments[var] in values:
                            return False
                        else:
                            values.add(assignments[var])

            elif constraint[0] == 'AllSame':
                values = set()
                for var in constraint[1:]:
                    if var in assignments:
                        values.add(assignments[var])

                if len(values) > 1:
                    return False
                
            elif constraint[0] == 'MaxCount':
                n = int(constraint[1])
                counts = {}
                for var in constraint[2:]:
                    if var in assignments:
                        if assignments[var] in counts:
                            counts[assignments[var]] += 1
                            if counts[assignments[var]] > n:
                                return False
                        else:
                            counts[assignments[var]] = 1

        return True

    def getDistance(self, state):
        '''Ignored. This is not an informed search problem.'''
        return 0

    def isGoal(self, state):
        '''
        The goal is reached when the agent is at the exit.
        
        Parameters:
            state ((int,int)): The state to check.

        Returns (bool): True if the given state is a goal state, False otherwise.
        '''        
        return len(state.unassignedVariables) == 0
    
    def forwardCheck(self, state, variable, value):
        '''
        Returns a new state with the given variable assigned the given value, and the domains of the other variables updated.

        Parameters:
            state ((assignments, unassignedVariables, domains)): The current state of the problem.
            variable (str): The variable to assign.
            value (Any): The value to assign to the variable.

        Returns (CSPState): A new state with the given variable assigned the given value, and the domains of the other variables updated.
        '''
        # Make deep copies of the state data.
        newAssignments = {k:v for k,v in state.assignments.items()}
        newAssignments[variable] = value
        newUnassignedVariables = [var for var in state.unassignedVariables if var != variable]
        newDomains = {k:v.copy() for k,v in state.domains.items()}
        
        # TODO: check each of the constraints, skipping those that don't involve
        # the variable being assigned. Handle each constraint type separately
        # (AllDiff, AllSame, MaxCount). For example, if the constraint is
        # AllDiff, remove the value assigned to the new variable from the
        # domains of the unassigned variables involved in the constraint.

        return CSPState(newAssignments, newUnassignedVariables, newDomains)


def main():
    if len(sys.argv) < 2:
        print("Usage: python3 csp.py <cspConfigFilename> [-fc] [-mrv] [-lcv]\n\n"+
              "Options:\n"+
                "  -fc: Use forward checking.\n"+
                "  -mrv: Use minimum remaining values ordering method.\n"+
                "  -lcv: Use least constraining value ordering method.\n")
        return
    
    forwardChecking = '-fc' in sys.argv
    minimumRemainingValues = '-mrv' in sys.argv
    leastConstrainingValue = '-lcv' in sys.argv

    print(f'Foward checking: {forwardChecking}')
    print(f'Minimum remaining values: {minimumRemainingValues}')
    print(f'Least constraining value: {leastConstrainingValue}')

    csp = CSP(sys.argv[1], forwardChecking, minimumRemainingValues, leastConstrainingValue)
    dfs = search.DFS(csp)

    node = dfs.nextState()
    while node is not None:
        if csp.isGoal(node.state):
            print(f'Search algorithm: {dfs.getName()}')
            print(f'States expanded: {dfs.getStatesExpanded()}')
            print(f'Max fringe size: {dfs.getMaxFringeSize()}')
            print(f'Solution path length: {len(node.pathActions)}')
            print(f'Solution path (actions): {", ".join(node.pathActions)}')
            break
        
        node = dfs.nextState()

    if node is None:
        print(f'Search algorithm: {dfs.getName()}')
        print(f'States expanded: {dfs.getStatesExpanded()}')
        print(f'Max fringe size: {dfs.getMaxFringeSize()}')
        print("No solution found.")



if __name__ == "__main__":
    main()